src/Sequence.cc

```diff
@@ -395,8 +423,12 @@ Sequence::execute_sequence_impl(Iterator step_begin, Iterator step_end, Context&
             ++step;
             continue;
         }
+
         if (comm and comm->immediate_termination_requested_)
-            throw Error{ gul14::cat(abort_marker, "Stop on user request") , step - steps_.begin() };
+        {
+            throw Error{ gul14::cat(abort_marker, "Stop on user request"),
+                         static_cast<StepIndex>(step - steps_.begin()) };
```

Comment:
Ah I had to understand this a bit more, that is because we do not set the `iterator_trains<Iterator>::difference_type` to be the same type as `StepIndex`.

Generally I'm not sure one should use casts when other solutions are possible, but (untested), what about

```c++
            throw Error{ gul14::cat(abort_marker, "Stop on user request"),
                         OptionalStepIndex{ step - steps_.begin() } };
```
or so?

Ah I see, the `gul14::optional` is of course broken. That works with `std::optional` but not with this self made thingy.

I still believe `ltaskolib` should use NOT use backports when the original stuff is existing in C++17.

Is there any reason we use `gul14::optional` at all?

We should create a GUL17, and use **that** here.

Reply:
> I still believe ltaskolib should use NOT use backports when the original stuff is existing in C++17.

I agree.

By the way, I got after pushing to my local `main` from the repo and a force rebase on my branch the following warning due to type narrowing:  

```bash
../src/Sequence.cc: In member function ‘task::Sequence::Iterator task::Sequence::execute_sequence_impl(task::Sequence::Iterator, task::Sequence::Iterator, task::Context&, task::CommChannel*)’:
../src/Sequence.cc:399:82: warning: narrowing conversion of ‘__gnu_cxx::operator-<task::Step*, std::vector<task::Step> >(step, ((task::Sequence*)this)->task::Sequence::steps_.std::vector<task::Step>::begin())’ from ‘__gnu_cxx::__normal_iterator<task::Step*, std::vector<task::Step> >::difference_type’ {aka ‘long int’} to ‘short unsigned int’ [-Wnarrowing]
  399 |             throw Error{ gul14::cat(abort_marker, "Stop on user request") , step - steps_.begin() };
      |                                                                             ~~~~~^~~~~~~~~~~~~~~~
[89/96] Linking target libtaskolib.so.0.4.0
```

One solution is to explicitly cast it too `StepIndex` but from my perspective it 'smells'.

Reply:
> By the way, I got after pushing to my local `main`

[...]

> One solution is to explicitly cast it too `StepIndex`

:thinking: .o( But that is exactly what Lars found and fixed in exactly that way?  )

![image](https://user-images.githubusercontent.com/16012374/208703359-9aca7246-fe89-4a86-8daf-63881d2dcb11.png)

Reply:
Why should this
```cpp
throw Error{ gul14::cat(abort_marker, "Stop on user request"),
                         OptionalStepIndex{ step - steps_.begin() } };
```
work? Isnt' that a narrowing conversion by definition?

I think the `static_cast` conveys exactly what we are doing. We are doing ~pointer~, ahem, iterator arithmetic, which results in too wide a type, but we know that the result must fit into a `StepIndex` because the `Sequence` class is made that way. I am open for cleaner solutions, of course... but I think the brace-initialization above *should* not work because it would be narrowing.

Reply:
> > I still believe ltaskolib should use NOT use backports when the original stuff is existing in C++17.
> 
> I agree.

I disagree. As long as we are using GUL14, we should stick to the types it provides for interoperability. Otherwise we'll end up with incompatibilities all over the place like when we used a few `std::string_view`s and could not pass them to GUL14's string functions.

Should we remove `string_view` and `optional` from future GUL17/GUL20? Ab-so-frickin'-lutely.

Reply:
> I am open for cleaner solutions, of course...

Well, clean would be to have `std::iterator_traits<Iterator>::difference_type` to return `StepIndex`, in which case I believe (untested) `std::difference(step, steps_.begin())` would return a `StepIndex`.

But that is even for me too much for too little, so I guess the `static cast` is ok here from my pov.